### ISSUE TITLE
Download speeds have decreased

### ISSUE THREAD

**Issue by hyuk199** [transmission/transmission#5952]
### What is the issue?

Since updating to 4.0.4, download speeds have decreased significantly.
I used to get 10+ MB downloads, but with 4.0.4 I'm getting less than 1 MB speed.
This is the overall speed, not for each file.
Please fix this issue.

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.4

**Comment by Pentaphon** [comment]
> I used to get 10+ MB downloads, but with 4.0.4 I'm getting less than 1 MB speed.

I don't get this issue at all. 4.0.4 seems a bit better than 4.0.3 which had more inconsistent speeds, but neither of them gave me 1MB speed. I'm using Windows and Linux though.

**Comment by jameshale** [comment]
I can confirm this.
Ventura 13.5.1 (22G90), Intel MacMini.
4.04 speeds atrocious. Not only are the speeds bad but connecting multiple torrents also has broken. I load up 4 torrents and only 2 connect and begin downloading (even if the tracking info shows 40 plus seeds). 
A quit and restart will sometimes bring the others up.
Speeds around 40 Kbs!!
Just reinstalled 4.03 and speeds up to 1 MBs per torrent. Something definitely changed in 4.04.

**Comment by stefanos82** [comment]
I have just tested it with the latest pull on my system, a GNU / Debian testing 64-bit, and while it was peaked at its highest download speed, for some reason something is dropping it down to less than 1 MB.

It behaved in a fluctuated way that tried to push to its highest speed available, only to get paused for a couple of seconds with an immediate drop to less or equal to 1MB on my side.

Weird behavior!

**Comment by thomasmichaelwallace** [comment]
Seeing similar here, on Mac, downgrading to 4.0.3 fixes.

Happy to supply any debug information if you let me know where to look 😄

**Comment by Quaap70** [comment]
Same issue, on Mac. Downgrading to 4.0.3 also fixes things here.

**Comment by MartinZubek** [comment]
Yup, I'm on the same boat, latest version is definitely worse with download speeds and connection in general.
I'm reverting to previous version for now.

**Comment by tearfur** [comment]
Sounds like all of the people having this problem (apart from one) are using the macOS app.

I'm gonna need help from people to test some test builds, since I don't have a Mac myself to test this. I'd like to know 2 things:

1. Do the latest nightly builds have this problem? You can find the .dmg files here: https://build.transmissionbt.com/job/trunk-mac/
2. Does this test build still have the same problem? How to install:
   1. Go to https://github.com/transmission/transmission/actions/runs/6045189233?pr=5954
   2. Under `Artifacts`, download `binaries-macos-11-from-tarball`. You will get `binaries-macos-11-from-tarball.zip`.
   3. Find `Applications/Transmission.app` inside the .zip and copy it to `/Applications` (I *think* this will work, honestly not sure how Mac works)

*Note: Installing the test builds will conflict with your existing Transmission installation.*

@Coeur @nevack @sweetppro Is there a better way to test this?

**Comment by GaryElshaw** [comment]
That's the best way.

**Comment by jameshale** [comment]
Just tried nightly
 4.0.3 (6b0e49bbb2)   - Transmission-fc4b7ed9f4.dmg

tested against 3 torrents

3 torrents after a three minute wait
T1 53 seeds - 5 of 5 peers-  2KB/s
T2 171 seeds - 6 of 6 peers-  340KB/s
T3 216 seeds - 10 of 11 peers-  1-3 MB/s

Then quit and went with the release of 4.03 to continue with same torrents
after a 1 minute wait
T1  now 6 of 6 peers - 300KB/s
T2 now 30 of 36 peers - 1.9MB/s
T3 now  39 of 60 peers- 1.56MB/s

The nightly speed is better overall than 4.04 but still down on 4.03 release also the number of peers on the nightly didn't come up to 4.03 values

**Comment by tearfur** [comment]
@jameshale Tranmission caches peers on shutdown to be used the next time it starts, so it is possible your 4.0.3 test got a headstart knowing a bunch of peers gathered by the nightly.

So could you please repeat the test by running 4.0.3 *first* then the nightly?

Also I'd appreciate if you could also test the test build I linked in (2), I've reverted the only macOS-specific commit between 4.0.3 and 4.0.4 to see if that was the cause.

**Comment by jameshale** [comment]
two torrents
release 4.03
T1 21 of 33 peers
T2 17 of 21 peer
quit and launch  nightly
T1 7 of 7 peers
T2 3 of 3 peers.
so there is something wrong
its almost 1am here so I will try the other build tomorrow.

**Comment by ckerr** [comment]
I'm always a little wary about slow xfer speed bug reports, since there are dozens of things that affect a swarm's speed. No two sessions are going to be alike, so it's usually difficult to tell if speed differences are due to the swarm or code changes. Still, since several people are reporting this, I took a look at [all the 4.0.3...4.0.4 diffs](https://github.com/transmission/transmission/compare/4.0.3...4.0.4) for likely suspects.

The only guess I have is this change in `libtransmission/bandwidth.cc` which was actually a copy-paste bugfix but maybe the old behavior had some benefit in priming the download pump more often?

```diff
diff --git a/libtransmission/bandwidth.cc b/libtransmission/bandwidth.cc
index cf0b88b98..d43242160 100644
--- a/libtransmission/bandwidth.cc
+++ b/libtransmission/bandwidth.cc
@@ -276,8 +276,8 @@ size_t tr_bandwidth::clamp(uint64_t now, tr_direction dir, size_t byte_count) co
                 now = tr_time_msec();
             }
 
-            auto const current = this->getRawSpeedBytesPerSecond(now, TR_DOWN);
-            auto const desired = this->getDesiredSpeedBytesPerSecond(TR_DOWN);
+            auto const current = this->getRawSpeedBytesPerSecond(now, dir);
+            auto const desired = this->getDesiredSpeedBytesPerSecond(dir);
             auto const r = desired >= 1 ? static_cast<double>(current) / desired : 0.0;
 
             if (r > 1.0)
```

If anyone experiencing this issue is able to rebuild from source code, please try reverting ^ that change and see if it does any good

**Comment by jameshale** [comment]
@tearfur Sorry, I am a GitHub novice so downloading the Mac-OS 11-from -tarball  is beyond me. Sure I see it listed but the link takes me to a page showing log output of a build and upload job. I do not actually see the tarball to download.

**Comment by GaryElshaw** [comment]
@jameshale you have to scroll to the bottom of the page, it's under 'Artifacts'.  https://github.com/transmission/transmission/actions/runs/6045189233?pr=5954

**Comment by wirezen** [comment]
I can confirm that 4.0.4 isn't working properly.
No more than one active downloading torrent at the time and the speed is decreased at ridiculous levels (less than 1MBps VS 30/40MBps).
I have just downgraded to 4.0.3 and everything works perfectly.

**Comment by cpd** [comment]
macOS Sonoma 14.0 Beta 7 (23A5337a)
Transmission 4.0.4 (24077e3511)
looks ok, i'm getting max speed

**Comment by Coeur** [comment]
Sorry, I don't have the bandwidth [pun intended] to test things at the moment.
And I wasn't quite using internet [!] in August, so when I came back from holidays, I did notice a behavioral regression in nightlies (#5943), and I attempted to pinpoint its origin (which I thought to be around 12-15 August), but those changes didn't go to the 4.0.x branch, so I may I have faced a different issue, or the regression is anterior to what I thought. Charles is pointing to a change from July 25 (#5821), so I can try to revert it and see if it helps with my magnets issue. @tearfur is your proposed fix from the branch "tearfur/on-port-duplicate-fix", or a different branch?

**Comment by jameshale** [comment]
@GaryElshaw  Ah found it. Unfortunately it refuses to open, even central-clicking and choosing open.

**Comment by tearfur** [comment]
> is your proposed fix from the branch "tearfur/on-port-duplicate-fix", or a different branch?

@Coeur You mean https://github.com/transmission/transmission/actions/runs/6045189233?pr=5954?

I don't have any proposed fix, I just reverted some suspected commits, then opened the PR to generate a test build using the CI.

**Comment by Coeur** [comment]
Ah, got it, it's "tearfur:macos-slow".

**Comment by GaryElshaw** [comment]
> <snip> Ah found it. Unfortunately it refuses to open, even central-clicking and choosing open.

Hmm, i think you might have to do some Terminal trickery on this if you feel up to it?

**Comment by dantearmok** [comment]
I've got way too many versions of Transmission installed...

Mac mini (late 2012, 2.5-GHz 4-Core i5), 16GB RAM, plus an assortment of USB & FW HDs(*).
macOS 13.6 Ventura (w/OCLP), PeakHour, PIA, and Xcode (all fully updated).
* Transmission uses a FW Drobo (long-term seeding) and three USB hard drives.

Pipe is 450 Mb wifi to Comcast coax 100/20.  And PIA is beaming me thru Toronto CA.

Transmission 4.0.3, 4.0.4, and the nightlies.  253 torrents.  Queues unchecked.  No upload or download limit.

Letter rip, tater chip!

After about 5 minutes my download is bouncing from 84 Mb to 95+.  The upload pegs 18 to 20 Mbps.  (Speeds per PeakHour, watching the wi-fi interface).

It is my impression that 4.0.4 took on the full load a tad faster than the others, fwiw.

By comparison: qB, under the same load, gets going quite a bit faster then does the same speeds.

**Comment by ckerr** [comment]
> By comparison: qB, under the same load, gets going quite a bit faster then does the same speeds.

I understand the impulse to compare to another client, but the question is about whether either https://github.com/transmission/transmission/issues/5952#issuecomment-1703862127 or https://github.com/transmission/transmission/issues/5952#issuecomment-1703399964 changes any behavior compared to `4.0.4` if anyone is up for doing A/B testing there? :pray:

**Comment by dantearmok** [comment]
I did A/B as it were.  I ran each of the mentioned apps, each with the same load.  .3 and .4 both pretty much pegged my bandwidth limits.  Is there a different load that you'd like to see?

**Comment by fixsssator** [comment]
Same issue on synology 920+ 
Download speeds have decreased

**Comment by odkrys** [comment]
I get a little better speed with 4.0.4.
It probably won't help solve the OP's problem, but anyway, in the openwrt arm armada 385 (armv7 1.6Ghz) environment, 4.0.3 usually reached speeds of 8~9MB/s and sometimes 10MB/s, but in 4.0.4, it usually reached 10~11MB/s and peaked up to 12MB/s.

**Comment by Awmusic12635** [comment]
I can also confirm I am seeing a similar slowdown on 4.0.4 compared to 4.0.3 in docker on ubuntu. Downgrading to 4.0.3 lead to immediate improvement in speeds.

**Comment by madpsy** [comment]
Also seeing this on Synology NAS. Downgrading to 4.0.3 fixes the issue right away. (uninstall then manual install using a package from https://synocommunity.com/package/transmission)

**Comment by tearfur** [comment]
@jameshale I might have a lead on what's going on for the decreased peer numbers. Would you please try disabling encryption to see if the numbers are back to normal?

**Comment by protectivedad** [comment]
I'm on Ubuntu Server and the speeds on 4.0.4 were ~300 KB/s. As soon as I downgraded to 4.0.3 it went to 10 MB/s about a x30 increase. I only mention it because of the comment that it was a macOS problem. That doesn't seem to be the case. Also I turned off encryption and there was no change to the speeds.

Just to clarify I'm using a linuxserver docker container. I run with 4.0.3 and it jumps to 4.3 MB/s immediately and goes up to 13 MB/s. I stop the container and start 4.0.4 a few seconds later the same download starts at ~300 KB/s and then flip between 85 KB/s and 1/2 MB/s.

**Comment by Coeur** [comment]
For those who can reproduce the issue, would you mind trying to build Transmission from source, and attempt to figure out which commit is guilty? There are 50 commits between the two releases: https://github.com/transmission/transmission/compare/4.0.3...4.0.4
That means just about ⌈log₂(50)⌉ = 6 attempts until you figure out precisely which one is faulty.

**Comment by tearfur** [comment]
> Also I turned off encryption and there was no change to the speeds.

Ahh... Then I'm out of ideas again.

**Comment by protectivedad** [comment]
> For those who can reproduce the issue, would you mind trying to build Transmission from source, and attempt to figure out which commit is guilty? There are 50 commits between the two releases: [4.0.3...4.0.4](https://github.com/transmission/transmission/compare/4.0.3...4.0.4) That means just about ⌈log₂(50)⌉ = 6 attempts until you figure out precisely which one is faulty.

I tried but the docker image uses Alpine Linux which uses musl. My dev environment is glibc.

You can probably reproduce the problem using a docker container.

**Comment by protectivedad** [comment]
> For those who can reproduce the issue, would you mind trying to build Transmission from source, and attempt to figure out which commit is guilty? There are 50 commits between the two releases: [4.0.3...4.0.4](https://github.com/transmission/transmission/compare/4.0.3...4.0.4) That means just about ⌈log₂(50)⌉ = 6 attempts until you figure out precisely which one is faulty.

Well I did it this way and found out it was commit 264746959 that was mentioned https://github.com/transmission/transmission/issues/5952#issuecomment-1703399964. Reverting the commit resolved the problem. I've put my short hand notes below.

```
https://releases.ubuntu.com/22.04/ubuntu-22.04.3-live-server-amd64.iso.torrent

Start with slowed down 4.0.4 image

Used original 4.0.4 binary:
Seeds 545 Peers 15
Speed max at 1.5 MB/s
Frequent disconnects from peers

Used newer 4.0.4 binary in older 4.0.3 image:
Seeds 738 Peers 18
Speed max 1.2 MB/s
Frequent disconnects from peers

Used older 4.0.3 binary in newer 4.0.4 image:
Seeds 776 Peers 17
Speed max 16.2 MB/s
Peer list is FAR more populated and the Down speeds are far higher

Used TEST 4.0.3 binary in newer 4.0.4 image:
Seeds 588 Peers 12
Speed max 14.7 MB/s
Peer list over 32 entries

Used TEST b86afe185 binary in newer 4.0.4 image:
Seeds 728 Peers 13
Speed Max 13.8 MB/s
Peer list over 32 entries

Used TEST 98f112c6b binary in newer 4.0.4 image:
Seeds 243 Peers 7
Speed Max 575 KB/s
Peer list 7 entries

Used TEST 2941005e3 binary in newer 4.0.4 image:
Seeds 257 Peers 7
Speed Max 12.9 MB/s
Peer list >32 entries

Used TEST 264746959 binary in newer 4.0.4 image:
Seeds 404 Peers 7
Speed Max 743 KB/s
Peer list 9 entries

Used TEST de26f38c7 binary in newer 4.0.4 image:
Seeds 555 Peers 11
Speed Max 12.4 MB/s
Peer list >32 entries

Used TEST 4.0.4 with 264746959 reverted binary in newer 4.0.4 image:
Seeds 113 Peers 3
Speed Max 14.6 MB/s
Peer list >32 entries
```

**Comment by tearfur** [comment]
@protectivedad Great work! Thank you very much.

Now the question is why did a supposed bug fix made things worse *only for some people*...

**Comment by wirezen** [comment]
> Now the question is why did a supposed bug fix made things worse _only for some people_...

i think it's worse for anyone but just some have directly seen the issue.

**Comment by tearfur** [comment]
@stefanos82 I'm not too sure how is `enum` Vs `enum class` relevant here. If you are running a debug build, you should already have crashed the moment you start to have network activity if `dir` was not `TR_DOWN` or `TR_UP`:

https://github.com/transmission/transmission/blob/6ead147620680259248bdebb764dae19f0cb8933/libtransmission/bandwidth.cc#L263-L265

https://github.com/transmission/transmission/blob/6ead147620680259248bdebb764dae19f0cb8933/libtransmission/transmission.h#L1621-L1624

**Comment by stefanos82** [comment]
...seriously, I have been reading `libtransmission/transmission.h` for hours and for some reason, I didn't see the very last lines with `tr_direction` validation! :face_exhaling: 

My sincere apologies if I cause any exasperation or discomfort, I feel so stupid right now :disappointed:

**Comment by tearfur** [comment]
@protectivedad and all others that are seeing this issue: What are your bandwidth limit settings? Do you have any bandwidth groups set? What are your Internet speeds?

For 264746959b853e243b7f2e4d52cffe5bbd5bdf79 to have any effect, you have to have a bandwidth limit set in some way or the other, and I'm trying to find a common pattern here.

**Comment by benjuang** [comment]
I figured it out.  Looks like if I have _anything_ set for my upload rate, it throttles to 0 KB/s up (or something about uploading breaks somewhere) and that seems to result in peers not sending me any data.  Here's a screen recording of not throttling my upload rate, then throttling upload rate.  I also try changing the upload rate.  Seems pretty consistent that once I check the "Upload rate" checkbox, my download speeds drop down to 1-2 MB/s.  (Which, coming from the dark ages of dial up, is still amazing, but I would like to have 10x that.)

https://github.com/transmission/transmission/assets/1890/15e1745a-03bd-4c60-8e59-b3657f6b9095

**Comment by tearfur** [comment]
@benjuang That's a fantastic step in the right direction!

I use the WebUI myself, and setting the upload limit works as expected, so it is very plausible this is a Mac UI bug. 

Could you also try enable the upload limit via WebUI instead of the Mac UI? I'll try look at the code as well.

CC @Coeur

**Comment by benjuang** [comment]
@tearfur Hm, nope, setting the upload limit via the WebUI also seems to result in decreased download speed and 0 KB/s up.  (I'm assuming by WebUI, you mean the remote access web interface.)

In case it matters, I'm on Transmission 4.0.4 (24077e3511) and MacOS X 13.5.

**Comment by tearfur** [comment]
@benjuang Yes, I do mean the remote access web interface.

On Linux and 4.0.4, I get around 40 kB/s when I set the upload limit to 50 kB/s, but not 0🤔

![image](https://github.com/transmission/transmission/assets/46261767/ad42f8a9-ebed-41fa-af76-cd40d4d4f2e2)

**Comment by EuroscanInstruments** [comment]
For info, if i set the upload rate to anything different from 0 or "unlimited", download is limited and very slow

**Comment by madpsy** [comment]
Can confirm the same behaviour of setting anything other than unlimited for upload causes this behaviour. Nothing Mac related here whatsoever - using the v4.0.4 Transmission package built for Synology.

**Comment by protectivedad** [comment]
As I pointed out it probably deals with the libc. It was definitely never a Mac only problem. I'm guessing Synology is not using GLIBC. Have any of the developers tried using a docker container to reproduce the problem as I suggested? Or try the problem in ArchLinux? They both use MUSL instead of GLIBC not sure about Mac but I'm guessing it doesn't use GLIBC either. If I'm off base sorry, but I haven't read anyone looking into this connection.

**Comment by tearfur** [comment]
OK, I can reproduce it by building from source in `alpine:latest`. I have not looked into why yet, but I'll be leaving my steps here if anyone wants to follow along.

<details>
<summary>Steps to build from source</summary>

1. In the host machine, `sudo docker run --rm --network host -it alpine:latest`.
2. In the container that spawned from the previous step, `apk add build-base git cmake openssl-dev curl-dev python3 ninja linux-headers gettext-dev`
3. `cd && git clone --recurse-submodules https://github.com/transmission/transmission.git && cd transmission/`
4. `cmake -B build_reldeb -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo`
5. `cd build_reldeb/ && cmake --build .`
6. `TRANSMISSION_WEB_HOME=../web/public_html/ daemon/transmission-daemon -f --log-level=info`

</details>

<details>
<summary>settings.json</summary>

```json
{
    "alt-speed-down": 50,
    "alt-speed-time-begin": 540,
    "alt-speed-time-day": 127,
    "alt-speed-time-enabled": false,
    "alt-speed-time-end": 1020,
    "alt-speed-up": 50,
    "announce-ip": "",
    "announce-ip-enabled": false,
    "anti-brute-force-enabled": false,
    "anti-brute-force-threshold": 100,
    "bind-address-ipv4": "0.0.0.0",
    "bind-address-ipv6": "::",
    "blocklist-enabled": false,
    "blocklist-url": "http://www.example.com/blocklist",
    "cache-size-mb": 4,
    "default-trackers": "",
    "dht-enabled": true,
    "download-dir": "/downloads",
    "download-queue-enabled": false,
    "download-queue-size": 5,
    "encryption": 1,
    "idle-seeding-limit": 30,
    "idle-seeding-limit-enabled": false,
    "incomplete-dir": "/downloads",
    "incomplete-dir-enabled": false,
    "lpd-enabled": true,
    "message-level": 4,
    "peer-congestion-algorithm": "",
    "peer-limit-global": 200,
    "peer-limit-per-torrent": 50,
    "peer-port": 51413,
    "peer-port-random-high": 65535,
    "peer-port-random-low": 49152,
    "peer-port-random-on-start": false,
    "peer-socket-tos": "le",
    "pex-enabled": true,
    "port-forwarding-enabled": true,
    "preallocation": 1,
    "preferred-transport": "tcp",
    "prefetch-enabled": true,
    "queue-stalled-enabled": true,
    "queue-stalled-minutes": 30,
    "ratio-limit": 2,
    "ratio-limit-enabled": false,
    "rename-partial-files": false,
    "rpc-authentication-required": false,
    "rpc-bind-address": "0.0.0.0",
    "rpc-enabled": true,
    "rpc-host-whitelist": "",
    "rpc-host-whitelist-enabled": false,
    "rpc-password": "{d67da72605661a02541a8aaf616df5c6e017bbbetXT5zoAX",
    "rpc-port": 9093,
    "rpc-socket-mode": "0750",
    "rpc-url": "/transmission/",
    "rpc-username": "",
    "rpc-whitelist": "",
    "rpc-whitelist-enabled": false,
    "scrape-paused-torrents-enabled": true,
    "script-torrent-added-enabled": false,
    "script-torrent-added-filename": "",
    "script-torrent-done-enabled": false,
    "script-torrent-done-filename": "",
    "script-torrent-done-seeding-enabled": false,
    "script-torrent-done-seeding-filename": "",
    "seed-queue-enabled": false,
    "seed-queue-size": 10,
    "speed-limit-down": 100,
    "speed-limit-down-enabled": false,
    "speed-limit-up": 50,
    "speed-limit-up-enabled": false,
    "start-added-torrents": true,
    "tcp-enabled": true,
    "torrent-added-verify-mode": "fast",
    "trash-original-torrent-files": false,
    "umask": "022",
    "upload-slots-per-torrent": 8,
    "utp-enabled": true
}
```

</details>

**Comment by tearfur** [comment]
Update:

I can reproduce this in a GLIBC environment now. Apparently this does not happen in `Debug` builds (or not obvious enough for me to notice). In `RelWithDebInfo` builds, the difference between setting and not setting an upload limit is very obvious.

**Comment by EuroscanInstruments** [comment]
Can you try and set an upload limit to 0 ?

**Comment by stefanos82** [comment]
Sorry for interrupting, but if it helps, I have noticed the following behavior last night: I opened `transmission-gtk` and attempted to download a simple magnet link for testing purposes and it would throttle down to ~5kb download speed, but as soon as I close it and opened `transmission-qt` to resume the download, it jumped straight to 2.4MB download speed.

I don't know whether it relates or not, but I wanted to report it.

Reminder: Can we please fix #5841 ? It prevents me from finding the actual commit for a bug report.

**Comment by tearfur** [comment]
@EuroscanInstruments @madpsy Can you guys try setting the upload limit to something big, such as 10 MB/s? That would be 10000 KB/s in Transmission's conventions.

I don't see the download speeds decreasing at this setting.

**Comment by EuroscanInstruments** [comment]
@tearfur With such a high upload limit, download speed is good and not decreasing.

**Comment by tearfur** [comment]
Then I think I know what's happening. Since the current speed limit includes protocol overhead in its calculation, an upload limit like 50 KB/s was just too low for Transmission to send out piece requests fast enough, so the other peers send out less piece data to us because they didn't get as much requests for them.

@ckerr I think it makes more sense to only calculate piece speed in `tr_bandwidth::clamp()`, wdyt?

**Comment by GaryElshaw** [comment]
Looking forward to finding out what the optimal ratio might be.

**Comment by tearfur** [comment]
@GaryElshaw IMO instead of a ratio of some sort, protocol overhead should not be involved in the calculation at all. Or else when the BT protocol itself is not at full capacity (like what's happening here), the file transfers can't operate at full speed even if you have a fast Internet.

**Comment by GaryElshaw** [comment]
> IMO instead of a ratio of some sort, protocol overhead should not be involved in the calculation at all. 

@tearfur I totally and completely agree with you, but if that isn't possible....

**Comment by tearfur** [comment]
My propesed change at #6082. I encourage everyone to test and tell us how well it worked for you.
If everyone's happy with it, then I'll go ahead and tidy it up before marking it as "ready to review".

### Linux

You'll just have to build from source.

### Windows and Mac

You can download test builds at https://github.com/transmission/transmission/actions/runs/6429223171?pr=6082 once they are ready.

For Windows, you'll get .zip file containing an installer, you can then install it just like any Transmission release.

For Mac, you'll get a .zip file containing a bunch of folders. Detailed instructions on how to run the test builds can be found at https://github.com/transmission/transmission/blob/main/docs/Testing-Transmission.md.

**Comment by andreygursky** [comment]
Maybe we should revert https://github.com/transmission/transmission/pull/5821? Since https://github.com/transmission/transmission/pull/6082 would revert commit https://github.com/transmission/transmission/commit/6e84fb89fa4ccc64894e8959c6e88c0d8432841a made on 02.03.2011, which didn't cause this issue here.

Even before the so called bug fix https://github.com/transmission/transmission/pull/5821 all those years it was possible to limit download and upload speeds separately. And the fine tune introduced in 2011 was not without a reason. Perhaps we could recall, why it was needed.

I know the hard limits of my internet connection and I just need to limit the transmission's **total** download and/or upload bandwidth usage (and I hope transmission to manage the limits in an optimal way). If I say e.g. 50 KB/s upload max, but transmission would now consume 150 KB/s (because of protocol overhead or really some bug causing excessive traffic), the limit I suppose to control wouldn't make so much sense anymore.

Sure, having the properly total limit an improvement would be to introduce an option to prioritize down or up traffic and a balance coefficient.

**Comment by tearfur** [comment]
@andreygursky I understand your concerns, but `tl;dr` I think #6082 will *most likely* restore the old behaviour for you for upload bandwidth.

> If I say e.g. 50 KB/s upload max, but transmission would now consume 150 KB/s (because of protocol overhead or really some bug causing excessive traffic)

If Transmission consumes up to 150 KB/s upload when you set the limit to 50 KB/s on #6082, then that means it always has been consuming up to 150 KB/s before #5821. This is because the bug existed right from the very start, and the corrections you mentioned never fixed/changed the part where the bug is.

Before #5821, assuming you never touched the default download bandwidth limits (i.e. limit disabled), and your upload limit is below that, then the code segment I removed in #6082 never had any effect on how much upload bandwidth Transmission consumes.

I happy to elaborate with references to the actual code if you are interested in knowing the details.

> all those years it was possible to limit download and upload speeds separately

You will be able to continue to do so, so no worries on that part.

**Comment by tearfur** [comment]
Another direction to resolve this issue:

1. The speed limit should include the protocol overhead, like it currently does in `4.0.4` and `main`.
2. Transmission should be changed report the total transfer speed (piece data + protocol overhead) to the user, instead of reporting just the piece data speed like it currently does.

This way, all users will need to slightly raise the upload speed limit to match the actual total transfer speed.

**Comment by GaryElshaw** [comment]
Our previous discussion. Users shouldn't have to blindly figure out for themselves what their optimal downstream is by being confused that it's effected by their upstream parameters.

I know how strange this might sound, but if we have found a serious issue with our application, maybe we could look at addressing it?

**Comment by protectivedad** [comment]
As a person who won't be doing any of the work I'm hesitant to comment, but I will anyway.

There is always a link between upload and download if for no other reason then there is a negotiation going on. That is the overhead and from my standpoint that should be a QOS issue with your router. The upload/download limits I feel should only apply to the actual data being transmitted and the negotiations should be separate from that. If the negotiations start to effect the user's internet connection then that is a problem for them to solve by using QOS on their router.

In this way any limiting of the upload won't effect the ability to find partners only the data that the partners send.

**Comment by andreygursky** [comment]
@tearfur, I agree, looking at #5821 I'd also say, yeah, looks like a typo, let's fix it. But if it will be verified that either reverting https://github.com/transmission/transmission/pull/5821 or applying https://github.com/transmission/transmission/pull/6082, then it would suggest this was actually intentional (or even if not it was not such harmful). I think a clean way would be to revert #5821. And then if really needed create a commit, which would explain what issue exactly it solves. Since #5821 doesn't mention any issue report. A new clean commit (after reverting #5821) might indeed refer to a potential issue you've mentioned above:
> If you had enabled the download limit, then the your upload speed was being shrunken down when your download speed is saturating the limit (which does not make any sense).

My intention is just not to mix up old bugs, new bugs, regressions, potential bug fixes, regression bug fixes, confirmed bug fixes even further. Since it would make tracking down new bugs/regression harder in future.

**Comment by tearfur** [comment]
> The speed limit should include the protocol overhead

Hmm, on more thought, this approach might be problematic after all.

Transmission can't see the TCP overhead, but it can see the uTP overhead. This means the "optimal" speed limit for the same Internet connection will be different depending on whether you are using TCP or uTP. Not to mention the vast majority of Transmissions probably are using a bit of both, so that complicates things further.

It is possible to ignore just the uTP overhead in calculation... But at that point, we might as well ignore all protocol overhead.

> So get_raw_speed_bytes_per_second() is not used now for displaying?

With a few exceptions, no, it is not.

The raw speed is only reported in the systemd journal and the gtk system tray icon tooltip, I'm not sure why is it the way it is...

**Comment by tearfur** [comment]
> BTW, this is only the behavior in recent main or was it also so before?

It tried to since the very beginning 6e84fb89fa4ccc64894e8959c6e88c0d8432841a, but it didn't even come close to working for upload bandwidth before #5821. Even if it did, I don't agree with the way it is implemented.

**Comment by mreid-tt** [comment]
I see this issue has been reported as completed. Any timeline for the next release to see the fix in action?

**Comment by tearfur** [comment]
The fix will be in the next release, `4.0.5` or `4.1.0`, whichever comes first.

As for when the next release will be, I'm afraid there isn't a concrete date as of now. See https://github.com/transmission/transmission/pull/6201#issuecomment-1792881611.

**Comment by ckerr** [comment]
@tearfur latest plan for 4.0.5 is https://github.com/transmission/transmission/pull/5263#issuecomment-1808443684
